sound/soc/intel/boards/sof_realtek_common.c

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include "../../codecs/rt1011.h"
+#include "sof_realtek_common.h"
+
+
+static const struct snd_soc_dapm_route rt1011_dapm_routes[] = {
+	/* speaker */
+	{ "Left Spk", NULL, "Left SPO" },
+	{ "Right Spk", NULL, "Right SPO" },
+};
+
+/*
+ * Make sure device's Unique ID follows this configuration:
+ *
+ * Two speakers:
+ *         0: left, 1: right
+ * Four speakers:
+ *         0: Woofer left, 1: Woofer right
+ *         2: Tweeter left, 3: Tweeter right
+ */
+static struct snd_soc_codec_conf rt1011_codec_confs[] = {
+	{
+		.dlc = COMP_CODEC_CONF(RT1011_DEV0_NAME),
+		.name_prefix = "Left",
+	},
+	{
+		.dlc = COMP_CODEC_CONF(RT1011_DEV1_NAME),
+		.name_prefix = "Right",
+	},
+};
+
+static struct snd_soc_dai_link_component rt1011_dai_link_components[] = {
+	{
+		.name = RT1011_DEV0_NAME,
+		.dai_name = RT1011_CODEC_DAI,
+	},
+	{
+		.name = RT1011_DEV1_NAME,
+		.dai_name = RT1011_CODEC_DAI,
+	},
+};
+
+static int rt1011_hw_params(struct snd_pcm_substream *substream,
+			    struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai;
+	int srate, i, ret = 0;
+
+	srate = params_rate(params);
+
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
+		/* 100 Fs to drive 24 bit data */
+		ret = snd_soc_dai_set_pll(codec_dai, 0, RT1011_PLL1_S_BCLK,
+					  100 * srate, 256 * srate);
+		if (ret < 0) {
+			dev_err(codec_dai->dev, "codec_dai pll not set\n");
+			return ret;
+		}
+
+		ret = snd_soc_dai_set_sysclk(codec_dai,
+					     RT1011_FS_SYS_PRE_S_PLL1,
+					     256 * srate, SND_SOC_CLOCK_IN);
+		if (ret < 0) {
+			dev_err(codec_dai->dev, "codec_dai sysclk not set\n");
+			return ret;
+		}
+
+		if (!strcmp(codec_dai->component->name, RT1011_DEV0_NAME)) {
```

Comment:
Will ``switch`` be better here? 
```
switch (i) {
        case 0: /* device 0 */
        tx_mask = 0x04;
        rx_mask = 0x01;
        ...
        break;
        case 1: /* device 1 */
        tx_mask = 0x08;
        rx_mask = 0x02;
        ...
        break;
        default:
        ...
        break;
}
ret = snd_soc_dai_set_tdm_slot(codec_dai, tx_mask,
                               rx_mask, 4, params_width(params));
if (ret < 0)
        break;
```

It can easily extend to supporting 4 speakers.

Reply:
or even make an array of mask structures, like
```
struct {
	unsigned int tx;
	unsigned int rx;
} mask[2] = {
	{.tx = 4, .rx = 1},
	{.tx = 8, .rx = 2},
};
...
ret = snd_soc_dai_set_tdm_slot(codec_dai, mask[i].tx,
                               mask[i].rx, 4, params_width(params));
```

Reply:
fixed